File: tests/test-net/src/main/java/pize/tests/net/ServerSide.java
package pize.tests.net;

import pize.net.security.KeyAES;
import pize.net.security.KeyRSA;
import pize.net.tcp.TcpConnection;
import pize.net.tcp.TcpListener;
import pize.net.tcp.TcpServer;
import pize.net.tcp.packet.PacketInfo;
import pize.net.tcp.packet.Packets;
import pize.tests.net.packet.EncodePacket;
import pize.tests.net.packet.MessagePacket;
import pize.tests.net.packet.PingPacket;

public class ServerSide implements TcpListener{
    
    public static void main(String[] args){
        new ServerSide();
    }
    
    
    private final TcpServer server;
    private final KeyRSA key;
    
    public ServerSide(){
        key = new KeyRSA(2048);
        
        server = new TcpServer(this);
        server.run("localhost", 5454);
        
        while(!Thread.interrupted());
        server.close();
    }
    
    
    @Override
    public void received(byte[] bytes, TcpConnection sender){
        try{
            final PacketInfo packetInfo = Packets.getPacketInfo(bytes);
            if(packetInfo == null){
                System.out.println("   received not packet");
                return;
            }
            
            switch(packetInfo.getPacketID()){
                case MessagePacket.PACKET_TYPE_ID -> {
                    final MessagePacket packet = packetInfo.readPacket(new MessagePacket());
                    System.out.println("   message: " + packet.getMessage());
                }
                
                case PingPacket.PACKET_TYPE_ID -> {
                    final PingPacket packet = packetInfo.readPacket(new PingPacket());
                    System.out.println("   client->server ping: " + (System.nanoTime() - packet.getTime()) / 1000000F);
                    packet.write(sender);
                }
                
                case EncodePacket.PACKET_TYPE_ID -> {
                    final EncodePacket packet = packetInfo.readPacket(new EncodePacket());
                    
                    final KeyAES clientKey = new KeyAES(key.decrypt(packet.getKey()));
                    System.out.println("   client's key received");
                    
                    sender.encode(clientKey);
                    System.out.println("   encoded with key (hash): " + clientKey.getKey().hashCode());
                }
            }
            
        }catch(Exception e){
            e.printStackTrace();
        }
    }
    
    @Override
    public void connected(TcpConnection connection){
        System.out.println("Client connected {");
        new EncodePacket(key.getPublic().getKey().getEncoded()).write(connection);
        System.out.println("   send public key");
    }
    
    @Override
    public void disconnected(TcpConnection connection){
        System.out.println("   Client disconnected\n}");
    }
    
}

File: tests/test-net/src/main/kotlin/pize/tests/net/ServerSide.kt
package pize.tests.net

import pize.net.security.KeyAES
import pize.net.security.KeyRSA
import pize.net.tcp.TcpConnection
import pize.net.tcp.TcpListener
import pize.net.tcp.TcpServer
import pize.net.tcp.packet.Packets.getPacketInfo
import pize.tests.net.packet.EncodePacket
import pize.tests.net.packet.MessagePacket
import pize.tests.net.packet.PingPacket

class ServerSide : TcpListener {
    private val server: TcpServer
    private val key: KeyRSA

    init {
        key = KeyRSA(2048)
        server = TcpServer(this)
        server.run("localhost", 5454)
        while (!Thread.interrupted());
        server.close()
    }

    override fun received(bytes: ByteArray?, sender: TcpConnection?) {
        try {
            val packetInfo = getPacketInfo(bytes!!)
            if (packetInfo == null) {
                println("   received not packet")
                return
            }
            when (packetInfo.packetID) {
                MessagePacket.Companion.PACKET_TYPE_ID -> {
                    val packet = packetInfo.readPacket(MessagePacket())
                    println("   message: " + packet.message)
                }

                PingPacket.Companion.PACKET_TYPE_ID -> {
                    val packet = packetInfo.readPacket(PingPacket())
                    println("   client->server ping: " + (System.nanoTime() - packet.time) / 1000000f)
                    packet.write(sender)
                }

                EncodePacket.Companion.PACKET_TYPE_ID -> {
                    val packet = packetInfo.readPacket(EncodePacket())
                    val clientKey = KeyAES(key.decrypt(packet.key))
                    println("   client's key received")
                    sender!!.encode(clientKey)
                    println("   encoded with key (hash): " + clientKey.key.hashCode())
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    override fun connected(connection: TcpConnection?) {
        println("Client connected {")
        EncodePacket(key.public!!.key!!.encoded).write(connection)
        println("   send public key")
    }

    override fun disconnected(connection: TcpConnection?) {
        println("   Client disconnected\n}")
    }

    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            ServerSide()
        }
    }
}
